protocols/bftcosi/bftcosi.go

```diff
@@ -228,125 +237,97 @@ func (bft *ProtocolBFTCoSi) RegisterOnSignatureDone(fn func(*BFTSignature)) {
+
+		var commitment abstract.Point
+		// store it and check if we have enough commitments
+		switch comm.TYPE {
+		case RoundPrepare:
+			bft.tempPrepareCommit = append(bft.tempPrepareCommit, comm.Commitment)
+			if len(bft.tempPrepareCommit) < len(bft.Children()) {
+				continue
+			}
+			commitment = bft.prepare.Commit(nil, bft.tempPrepareCommit)
+			if bft.IsRoot() {
+				if err := bft.startChallenge(RoundPrepare); err != nil {
+					return nil
+				}
+				continue
+			}
+		case RoundCommit:
+			bft.tempCommitCommit = append(bft.tempCommitCommit, comm.Commitment)
+			if len(bft.tempCommitCommit) < len(bft.Children()) {
+				continue
+			}
+			commitment = bft.commit.Commit(nil, bft.tempCommitCommit)
+			if bft.IsRoot() {
+				// do nothing:
+				// stop the processing of the round, wait the end of
+				// the "prepare" round: calls startChallengeCommit
+				continue
+			}
+
+			log.Lvl4(bft.Name(), "BFTCoSi handle Commit COMMIT")
 		}
-		commitment = bft.commit.Commit(nil, bft.tempCommitCommit)
-		bft.commitCommitDone <- true
-		if bft.IsRoot() {
-			// do nothing:
-			// stop the processing of the round, wait the end of
-			// the "prepare" round: calls startChallengeCommit
-			return nil
+		// set same RoundType as for the received commitment:
+		typedCommitment := &Commitment{
+			TYPE:       comm.TYPE,
+			Commitment: commitment,
 		}
-
-		log.Lvl4(bft.Name(), "BFTCoSi handle Commit COMMIT")
-	}
-	// set same RoundType as for the received commitment:
-	typedCommitment := &Commitment{
-		TYPE:       comm.TYPE,
-		Commitment: commitment,
-	}
-	return bft.SendToParent(typedCommitment)
-}
-
-// startChallenge creates the challenge and sends it to its children
-func (bft *ProtocolBFTCoSi) startChallenge(t RoundType) error {
-
-	log.Lvl4(bft.Name(), "RoundType:", t)
-	if t == RoundPrepare {
-		// need to hash the message before so challenge in both phases are not
-		// the same
-		data := sha512.Sum512(bft.Msg)
-		ch, err := bft.prepare.CreateChallenge(data[:])
-		if err != nil {
+		if err := bft.SendToParent(typedCommitment); err != nil {
 			return err
 		}
-		bftChal := &ChallengePrepare{
-			Challenge: ch,
-			Msg:       bft.Msg,
-			Data:      bft.Data,
-		}
-
-		return bft.handleChallengePrepare(bftChal)
-	}
-
-	// make sure the Announce->Commit has been done for the commit phase
-	<-bft.commitCommitDone
-	// commit phase
-	ch, err := bft.commit.CreateChallenge(bft.Msg)
-	if err != nil {
-		return err
 	}
-
-	// send challenge + signature
-	cc := &ChallengeCommit{
-		Challenge: ch,
-		Signature: &BFTSignature{
-			Msg:        bft.Msg,
-			Sig:        bft.prepareSignature,
-			Exceptions: bft.tempExceptions,
-		},
-	}
-	return bft.handleChallengeCommit(cc)
+	return nil
 }
 
 // handleChallengePrepare collects the challenge-messages
-func (bft *ProtocolBFTCoSi) handleChallengePrepare(ch *ChallengePrepare) error {
+func (bft *ProtocolBFTCoSi) handleChallengePrepare(msg challengePrepareChan) error {
+	log.Lvl3(bft.Name(), "RoundType: 0")
```

Comment:
Again compatibility with `log.Lvl3(ann.TYPE)` which prints `0` and `1` instead of `Prepare` and `Commit`. Furthermore it's more easy to parse `0` and `1` in the brain than `Prepare` and `Commit` when going through logs with a lot of lines ;)

But now that it works I'll take away those messages (hoping I won't need them soon again ;)